### ISSUE TITLE
Segmentstore is showing IllegalContainerStateException for a particular container multiple time while starting container

### ISSUE THREAD

**Issue by Ranjan-Padhi** [pravega/pravega#3751]
observed IllegalContainerStateException multiple times for same container in segmentstore while starting container during container changes 

After analyzed the log we found while staring container `i.p.s.s.h.ZKSegmentContainerMonitor - Starting Container 2` segmentstore is throwing `IllegalContainerStateException: Container 2 is in an invalid state` multiple times and after 1 min same container recover `i.p.s.s.c.StreamSegmentContainer - SegmentContainer[2]: Started.` and works properly.

Again after 9-10 mins same IllegalContainerStateException coming for same container(container 2) and recovered after some, like that multiple times same exception is coming and recovering

As per Pravega configuration 32 containers configured
`pravegaservice.containerCount: "32"`

Same issue observed in multiple setup for different different container.

Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```

Deployment Details:-
```
bookkeeper:0.5.0-2236.5228e2d
pravega:0.5.0-2236.5228e2d
zookeeper:3.5.4-beta-operator
pravega-operator:0.3.2
zookeeper-operator:0.2.1
```

Segmentstore log :-

```
2019-05-02 16:02:07,289 29702290 [core-15] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [18, 2, 3, 4, 20, 6, 12, 13], Current = [18, 3, 4, 20, 6, 12, 13], PendingTasks = [], ToStart = [2], ToStop = [].
2019-05-02 16:02:07,289 29702290 [core-15] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Starting Container 2.
2019-05-02 16:02:07,289 29702290 [core-15] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_2]: Created empty database directory '/tmp/pravega/cache/cache_Container_2'.
2019-05-02 16:02:07,440 29702441 [core-15] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[2]: Enter RecoveryMode.
2019-05-02 16:02:07,440 29702441 [core-15] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[2]: Reset.
2019-05-02 16:02:07,440 29702441 [core-15] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[2] Enter RecoveryMode.
2019-05-02 16:02:07,440 29702441 [core-15] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[2]: Cleared.
2019-05-02 16:02:12,326 29707327 [epollEventLoopGroup-11-8] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: 47f840ce-11dc-465d-af3e-8e8f9c70b28a on segment: longevity/smallKey/283.#epoch.177
2019-05-02 16:02:12,326 29707327 [epollEventLoopGroup-11-8] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
io.pravega.segmentstore.server.IllegalContainerStateException: Container 2 is in an invalid state for this operation. Expected: RUNNING; Actual: STARTING.
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.ensureRunning(StreamSegmentContainer.java:769)
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.getAttributes(StreamSegmentContainer.java:383)
        at io.pravega.segmentstore.server.store.StreamSegmentService.lambda$getAttributes$3(StreamSegmentService.java:75)
        at io.pravega.segmentstore.server.store.SegmentContainerCollection.invoke(SegmentContainerCollection.java:72)
        at io.pravega.segmentstore.server.store.StreamSegmentService.getAttributes(StreamSegmentService.java:73)
        at io.pravega.segmentstore.server.host.handler.AppendProcessor.setupAppend(AppendProcessor.java:164)
        at io.pravega.shared.protocol.netty.WireCommands$SetupAppend.process(WireCommands.java:501)
        at io.pravega.segmentstore.server.host.handler.ServerConnectionInboundHandler.channelRead(ServerConnectionInboundHandler.java:46)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)

**Same exception is coming multiple times then recovering** 
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[2]: MapStreamSegment SegmentId = 67583, Name = 'longevity/smallKey/283.#epoch.177', Active = 4
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[2]: MapStreamSegment SegmentId = 67598, Name = 'longevity/smallKey/282.#epoch.176', Active = 5
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[2] Recovery completed. Epoch = 12, Items Recovered = 69539, Time = 35804ms.
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.s.c.StreamSegmentContainerMetadata - SegmentContainer[2]: Exit RecoveryMode.
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.s.reading.ContainerReadIndex - ReadIndex[2] Exit RecoveryMode.
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.c.c.AbstractThreadPoolService - OperationProcessor[2]: Started.
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.server.logs.DurableLog - DurableLog[2]: Started (Online).
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.s.s.c.TableMetadataStore - _system/containers/metadata_2: Metadata Segment pinned. Name = '1', Id = '{}'
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.c.c.AbstractThreadPoolService - SegmentContainer[2]: Started.
2019-05-02 16:02:43,245 29738246 [core-15] INFO  i.p.c.c.AbstractThreadPoolService - StorageWriter[2]: Started.
2019-05-02 16:02:43,245 29738246 [core-17] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container 2 has been registered.
2019-05-02 16:02:43,245 29738246 [core-24] INFO  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[2]: Started.
2019-05-02 16:02:47,421 29742422 [core-6] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [18, 2, 3, 4, 20, 6, 12, 13], Current = [18, 2, 3, 4, 20, 6, 12, 13], PendingTasks = [], ToStart = [], ToStop = [].
```
Again after 10 min same IllegalContainerStateException coming for same container

```
2019-05-02 16:13:37,446 30392447 [core-13] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [18, 2, 3, 4, 20, 6, 12, 13], Current = [18, 3, 4, 20, 6, 12, 13], PendingTasks = [], ToStart = [2], ToStop = [].
2019-05-02 16:13:37,446 30392447 [core-13] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Starting Container 2.
2019-05-02 16:13:37,447 30392448 [core-13] INFO  i.p.s.s.impl.rocksdb.RocksDBCache - RocksDBCache[Container_2]: Created empty database directory '/tmp/pravega/cache/cache_Container_2'.
2019-05-02 16:13:45,249 30400250 [core-8] INFO  i.p.s.s.l.ContainerMetadataUpdateTransaction - MetadataUpdate[2-0]: Recovering MetadataCheckpointOperation with SequenceNumber 177361.
2019-05-02 16:13:45,249 30400250 [core-8] INFO  i.p.s.server.CacheManager - CacheManager Registered ReadIndex[2-67672] (longevity/smallKey/292.#epoch.182).
2019-05-02 16:13:47,333 30402334 [epollEventLoopGroup-11-7] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: ee5eeb4e-2972-49de-b020-f9c999a0607a on segment: longevity/smallKey/292.#epoch.182
2019-05-02 16:13:47,334 30402335 [epollEventLoopGroup-11-7] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
io.pravega.segmentstore.server.IllegalContainerStateException: Container 2 is in an invalid state for this operation. Expected: RUNNING; Actual: STARTING.
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.ensureRunning(StreamSegmentContainer.java:769)
        at io.pravega.segmentstore.server.containers.StreamSegmentContainer.getAttributes(StreamSegmentContainer.java:383)
        at io.pravega.segmentstore.server.store.StreamSegmentService.lambda$getAttributes$3(StreamSegmentService.java:75)
        at io.pravega.segmentstore.server.store.SegmentContainerCollection.invoke(SegmentContainerCollection.java:72)
        at io.pravega.segmentstore.server.store.StreamSegmentService.getAttributes(StreamSegmentService.java:73)
        at io.pravega.segmentstore.server.host.handler.AppendProcessor.setupAppend(AppendProcessor.java:164)
        at io.pravega.shared.protocol.netty.WireCommands$SetupAppend.process(WireCommands.java:501)
        at io.pravega.segmentstore.server.host.handler.ServerConnectionInboundHandler.channelRead(ServerConnectionInboundHandler.java:46)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
```

**Comment by Ranjan-Padhi** [comment]
[pravega-pravega-segmentstore-1.zip](https://github.com/pravega/pravega/files/3146795/pravega-pravega-segmentstore-1.zip)

**Comment by sumit-bm** [comment]
Does segment container frequent restart normal?

**Comment by shrids** [comment]
The segment container 2 stopped due to the error.
```
2019-05-02 16:02:01,956 29696957 [core-13] ERROR i.p.s.s.s.StreamSegmentContainerRegistry - Critical failure for SegmentContainer Container Id = 2, State = FAILED. {}
io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 67672, Offsets = [1491616209-1558725073), SeqNo = 177433' because of data differences at SegmentOffset 1552433617. Segment = longevity/smallKey/292.#epoch.182, Storage: Length=1558725073(Not Sealed), Metadata: Length=1491616209(Not Sealed)
	at io.pravega.segmentstore.server.writer.SegmentAggregator.verifySame(SegmentAggregator.java:1391)
	at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$reconcileData$54(SegmentAggregator.java:1381)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by andreipaduroiu** [comment]
This issue is not due to the container restarts. It's due to the Controller not waiting for the Segment Container to fully start before issuing requests to it. If you read the full error message, it says that it is rejecting the request due to the Segment Container currently Starting (and not Running).

**Comment by shrids** [comment]
Thanks @andrei. Went through the logs again and please find my findings below. 
If the container restart observed in the logs is a non-issue then we should close this issue.

**Details:**
a. Segment container 2 has always been assigned to this SegmentStore by the controller. From the logs we can see that 
    ` 2019-05-02 14:07:27,043` to `2019-05-03 05:58:19,365` Segment container 2 is assigned to this SSS instance.

b. The logs indicate that segment container 2 stopped twice in this run.
```
Line 480645: 2019-05-02 16:02:01,155 29696156 [core-10] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[2]: Writer failed. Shutting down StreamSegmentContainer.
Line 483141: 2019-05-02 16:13:28,288 30383289 [core-22] WARN  i.p.s.s.c.StreamSegmentContainer - SegmentContainer[2]: Writer failed. Shutting down StreamSegmentContainer.
```
Both these times the segment container restarted and came back up.

c. The probable reason for the segment container going down is 
```
// first instance
2019-05-02 16:02:01,152 29696153 [core-14] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[37750].CriticalError.
io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 67672, Offsets = [1491616209-1558725073), SeqNo = 177433' because of data differences at SegmentOffset 1552433617. Segment = longevity/smallKey/292.#epoch.182, Storage: Length=1558725073(Not Sealed), Metadata: Length=1491616209(Not Sealed)
	at io.pravega.segmentstore.server.writer.SegmentAggregator.verifySame(SegmentAggregator.java:1391)
	at io.pravega.segmentstore.server.writer.SegmentAggregator.lambda$reconcileData$54(SegmentAggregator.java:1381)
	
// second instance
2019-05-02 16:13:28,286 30383287 [core-9] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[15].CriticalError.
io.pravega.segmentstore.server.writer.ReconciliationFailureException: Unable to reconcile operation 'AggregatedAppend: SegmentId = 67672, Offsets = [1625833937-1692942801), SeqNo = 185709' because of data differences at SegmentOffset 1653096913. Segment = longevity/smallKey/292.#epoch.182, Storage: Length=1656242641(Not Sealed), Metadata: Length=1625833937(Not Sealed)
	at io.pravega.segmentstore.server.writer.SegmentAggregator.verifySame(SegmentAggregator.java:1391)

```

d. Also observed the following logs around the time the segment container stopped
```
2019-05-02 16:01:45,448 29680449 [storage-io-49] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[37748].Error.
java.io.IOException: I/O error
	at sun.nio.ch.FileDispatcherImpl.force0(Native Method)
	at sun.nio.ch.FileDispatcherImpl.force(FileDispatcherImpl.java:76)
	at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:388)
	at io.pravega.storage.filesystem.FileSystemStorage.doWrite(FileSystemStorage.java:323)

2019-05-02 16:13:14,717 30369718 [storage-io-59] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[13].Error.
java.io.IOException: I/O error
	at sun.nio.ch.FileDispatcherImpl.force0(Native Method)
	at sun.nio.ch.FileDispatcherImpl.force(FileDispatcherImpl.java:76)
	at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:388)
	at io.pravega.storage.filesystem.FileSystemStorage.doWrite(FileSystemStorage.java:323)
```

**Comment by andreipaduroiu** [comment]
Those are addressed in other issues (#3747, #3646). This one was for the WARN about IllegalContainerStateException while starting.

**Comment by shrids** [comment]
Thanks @andreipaduroiu, closing this issue based on the comment as WARN log message with `IllegalContainerStateException` when the container is starting is expected.
